src/txrequest.h

```diff
@@ -0,0 +1,332 @@
+    //! CANDIDATE_BEST. Returns whether the Entry still exists.
+    bool MakeCompleted(typename Index::index<ByTxid>::type::iterator it);
+
+    //! Make the data structure consistent with a given point in time:
+    //! - REQUESTED entries with exptime <= now are turned into COMPLETED.
+    //! - CANDIDATE_NEW entries with reqtime <= now are turned into CANDIDATE_{OTHER,BEST}.
+    //! - CANDIDATE_{OTHER,BEST} entries with reqtime > now are turned into CANDIDATE_NEW.
+    void SetTimePoint(std::chrono::microseconds now);
+
+public:
+    //! Construct a TxRequestTracker with the specified parameters.
+    TxRequestTracker(std::chrono::microseconds inbound_delay, std::chrono::microseconds timeout,
+        bool deterministic = false);
+
+    // Disable copying and assigning (a default copy won't work due the stateful EntryTxidExtractor).
+    TxRequestTracker(const TxRequestTracker&) = delete;
+    TxRequestTracker& operator=(const TxRequestTracker&) = delete;
+
+    //! A peer went offline, delete any data related to it.
+    void DeletedPeer(uint64_t uint64_t);
+
+    //! For whatever reason, we no longer need this txid. Delete any data related to it.
+    void AlreadyHaveTx(const uint256& txid);
+
+    //! We received a new inv, enter it into the data structure.
+    void ReceivedInv(uint64_t peer, bool outbound, const uint256& txid, std::chrono::microseconds now);
+
+    //! Find the txids to request now from peer.
+    std::vector<uint256> GetRequestable(uint64_t peer, std::chrono::microseconds now);
+
+    //! Inform the data structure that a txid was requested. This can only be called for txids returned by the last
+    //! GetRequestable call (which must have been for the same peer), with only other RequestedTx and AlreadyHaveTx
+    //! calls in between (which must have been for the same peer but different txids).
+    void RequestedTx(uint64_t peer, const uint256& txid, std::chrono::microseconds now);
+
+    //! We received a response (a tx, or a NOTFOUND) for txid from peer. Note that if a good tx is received (such
+    //! that we don't need it anymore), AlreadyHaveTx should be called instead of (or in addition to)
+    //! ReceivedResponse.
+    void ReceivedResponse(uint64_t peer, const uint256& txid);
+
+    //! Count how many in-flight transactions a peer has.
+    size_t CountInFlight(uint64_t peer) const;
+
+    //! Count how many transactions are being tracked for a peer (including timed-out ones and in-flight ones).
+    size_t CountTracked(uint64_t peer) const;
+
+    //! Count how many announcements are being tracked in total across all peers and transactions.
+    size_t Size() const { return m_index.size(); }
+
+    //! Query configuration parameter inbound delay.
+    std::chrono::microseconds GetInboundDelay() const { return m_inbound_delay; }
+
+    //! Query configuration parameter timeout.
+    std::chrono::microseconds GetTimeout() const { return m_timeout; }
+};
+
+#endif
```

Comment:
3d5b4f20015246b7a0dd74b2dc89f389204dd9b8

```suggestion
#endif // BITCOIN_TXREQUEST_H
```

See: [Source code organization](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#source-code-organization)

Reply:
Done.